File: src/sicav_checker/api/routes/verification.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query

from sicav_checker.api.dependencies import get_project_service
from sicav_checker.api.schemas.api_models import PaginatedAnomalies, VerificationRunResponse, VerificationSummary
from sicav_checker.services.project_service import ProjectService

router = APIRouter(prefix="/projects/{project_id}/verification", tags=["verification"])


@router.post("/run", response_model=VerificationRunResponse)
def run_verification(project_id: str, service: ProjectService = Depends(get_project_service)) -> VerificationRunResponse:
    run = service.run_verification(project_id)
    return VerificationRunResponse(run_id=run.id, status=run.status)


@router.get("/{run_id}", response_model=VerificationSummary)
def get_verification(project_id: str, run_id: str, service: ProjectService = Depends(get_project_service)) -> VerificationSummary:
    run = service.get_run(project_id, run_id)
    return VerificationSummary(run_id=run.id, status=run.status, **run.summary)


@router.get("/{run_id}/anomalies", response_model=PaginatedAnomalies)
def get_anomalies(project_id: str, run_id: str, page: int = Query(1, ge=1), page_size: int = Query(50, ge=1, le=500), service: ProjectService = Depends(get_project_service)) -> PaginatedAnomalies:
    run = service.get_run(project_id, run_id)
    start = (page - 1) * page_size
    end = start + page_size
    return PaginatedAnomalies(items=run.anomalies[start:end], total=len(run.anomalies), page=page, page_size=page_size)

File: frontend/src/api/verification.ts
import { api } from './client';
import type { ComparisonEvidence, DocumentPageResponse, EvidenceListResponse, PaginatedAnomalies, VerificationRunRequest, VerificationRunResponse, VerificationSummary } from '../types/api';

export async function runVerification(projectId: string, payload?: VerificationRunRequest) {
  const { data } = await api.post<VerificationRunResponse>('/projects/' + projectId + '/verification/run', payload ?? {});
  return data;
}

export async function getVerification(projectId: string, runId: string) {
  const { data } = await api.get<VerificationSummary>('/projects/' + projectId + '/verification/' + runId);
  return data;
}

export async function getAnomalies(projectId: string, runId: string, page = 1, pageSize = 25) {
  const { data } = await api.get<PaginatedAnomalies>('/projects/' + projectId + '/verification/' + runId + '/anomalies', {
    params: { page, page_size: pageSize },
  });
  return data;
}

export async function getEvidenceList(projectId: string, runId: string) {
  const { data } = await api.get<EvidenceListResponse>('/projects/' + projectId + '/verification/' + runId + '/evidence');
  return data;
}

export async function getEvidenceDetail(projectId: string, runId: string, evidenceId: string) {
  const { data } = await api.get<ComparisonEvidence>('/projects/' + projectId + '/verification/' + runId + '/evidence/' + encodeURIComponent(evidenceId));
  return data;
}

export async function getDocumentPage(projectId: string, documentId: string, pageNumber: number, runId?: string, evidenceId?: string) {
  const { data } = await api.get<DocumentPageResponse>('/projects/' + projectId + '/documents/' + encodeURIComponent(documentId) + '/page/' + pageNumber, {
    params: { run_id: runId, evidence_id: evidenceId },
  });
  return data;
}

export async function listProjectRuns(projectId: string) {
  const { data } = await api.get<Array<{ run_id: string; status: string; created_at: string; summary: Record<string, unknown> }>>('/projects/' + projectId + '/runs');
  return data;
}
